1plusX RTD submodule: New RTD Module

bwhisp opened the pull request:
<!--
Thank you for your pull request. Please make sure this PR is scoped to one change, and that any added or changed code includes tests with greater than 80% code coverage. See https://github.com/prebid/Prebid.js/blob/master/CONTRIBUTING.md#testing-prebidjs for documentation on testing Prebid.js.
-->

## Type of change
<!-- Remove items that don't apply and/or select an item by changing [ ] to [x] -->
- [ ] Bugfix
- [x] Feature (New RTD submodule)
- [ ] New bidder adapter  <!--  IMPORTANT: if checking here, also submit your bidder params documentation here https://github.com/prebid/prebid.github.io/tree/master/dev-docs/bidders --> 
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Does this change affect user-facing APIs or examples documented on http://prebid.org?
- [ ] Other

## Description of change
<!-- Describe the change proposed in this pull request -->
This PR adds a RTD submodule for 1plusX 
This module will fetch first party data from 1plusX's Profile API consisting of : 
  - the audiences the user is part of 
  - the topics related to the page
 
Then it will transmit this data to the supported & selected bid adapters through `setBidderConfig`

Contents of the PR : 
    - 1plusX RTD module: `modules/1plusXRtdProvider.js`
    - Unit tests for 1plusX RTD module: `test/spec/modules/1plusXRtdProvider_spec.js`
    - Integration example for 1plusX RTD module: `integrationExamples/gpt/1plusXRtdProviderExample.html`
    - Documentation for 1plusX RTD module: `modules/1plusXRtdProvider.md`

For any changes that affect user-facing APIs or example code documented on http://prebid.org, please provide:

- PR adding the documentation : prebid/prebid.github.io#3873

## Other information
I am not 100% sure about where to write the data in the ORTB2 format when updating the bid adapter configs.

The data I am fetching from 1plusX's Profile API consists of lists of custom-defined Ids that don't follow the standard taxonomy. 
While I found info about where to write audience segments & topics in ORTB2; it seemed that it concerned only those following the standard taxonomy. 

I reached a solution by combining the [ORTB2 specs](https://www.iab.com/wp-content/uploads/2016/03/OpenRTB-API-Specification-Version-2-5-FINAL.pdf) (pages 31 & 25) AND some examples from other RTD Providers : 
   - audience segments will be written in an element of `ortb2.user.data` array
   - topics will be written in `ortb2.site.keywords`

I will add comments to the related parts of the code, please review these parts with special care.

bwhisp commented on modules/1plusXRtdProvider.js:
```diff
@@ -0,0 +1,211 @@
+ * @param {string[]} topics Represents the topics of the page
+ * @returns {Object} Object describing the updates to make on bidder configs
+ */
+export const buildOrtb2Updates = ({ segments = [], topics = [] }) => {
```
This function will create the updates to the ORTB2 object to be set on bid adapters configs

bwhisp commented on modules/1plusXRtdProvider.js:
```diff
@@ -0,0 +1,211 @@
+ * @returns {Object} Object describing the updates to make on bidder configs
+ */
+export const buildOrtb2Updates = ({ segments = [], topics = [] }) => {
+  const userData = {
+    name: '1plusX.com',
+    segment: segments.map((segmentId) => ({ id: segmentId }))
+  };
```
Here is the element I will add to `ortb2.user.data` containing the audience segments

bwhisp commented on modules/1plusXRtdProvider.js:
```diff
@@ -0,0 +1,211 @@
+  const { site, userData } = ortb2Updates;
+  const bidderConfigCopy = mergeDeep({}, bidderConfigs[bidder]);
+
+  const currentSite = deepAccess(bidderConfigCopy, 'ortb2.site')
+  const updatedSite = mergeDeep(currentSite, site);
```
Writing topics to `ortb2.site.keywords`

patmmccann replied:
this is the wrong place for these data

bwhisp commented on modules/1plusXRtdProvider.js:
```diff
@@ -0,0 +1,211 @@
+  const currentSite = deepAccess(bidderConfigCopy, 'ortb2.site')
+  const updatedSite = mergeDeep(currentSite, site);
+
+  const currentUserData = deepAccess(bidderConfigCopy, 'ortb2.user.data') || [];
+  const updatedUserData = [
+    ...currentUserData.filter(({ name }) => name != userData.name),
+    userData
+  ];
```
In `ortb2.user.data`; replacing the element bearing `1plusX.com` name with the element defined at `buildOrtb2Updates::const userData`

patmmccann replied:
it is confusing you say you only support AppNexus and Magnite, yet you seem to be writing to an accessible object for all bidders here.

Fawke reviewed (commented):
Hi @bwhisp,

Thanks for the PR. While I'm still reviewing your changes, here are a few quick things that I noticed.

- Your test coverage is at ~67%. We need to have more than 80% code coverage.
- You need to add your module in `.submodules.json` file.

**to check code-coverage, you can run the command, `gulp test-coverage` and see the report. Here's a screenshot of what I got when I ran it.

<img width="1663" alt="Screenshot 2022-07-05 at 4 29 38 PM" src="https://user-images.githubusercontent.com/14229985/177344594-6a6022d0-1f26-4708-ad0c-c042c2d0efd3.png">

patmmccann commented on modules/1plusXRtdProvider.md:
```diff
@@ -0,0 +1,76 @@
+| params.timeout    | Integer       | timeout (ms)                                                     | 1000ms                                                   |
+
+## Supported Bidders
+At the moment only Appnexus (`appnexus`) and Magnite (`rubicon`) are supported
```
This is a rules violation

bwhisp replied:
I just removed the `SUPPORTED_BIDDERS` limitation

bwhisp commented:
> Hi @bwhisp,
> 
> Thanks for the PR. While I'm still reviewing your changes, here are a few quick things that I noticed.
> 
> * Your test coverage is at ~67%. We need to have more than 80% code coverage.
> * You need to add your module in `.submodules.json` file.
> 
> **to check code-coverage, you can run the command, `gulp test-coverage` and see the report. Here's a screenshot of what I got when I ran it.
> 
> <img alt="Screenshot 2022-07-05 at 4 29 38 PM" width="1663" src="https://user-images.githubusercontent.com/14229985/177344594-6a6022d0-1f26-4708-ad0c-c042c2d0efd3.png">

Hi @Fawke 

Thank you for the heads up!  
I updated my tests for better coverage, adding tests for the `extractConfig` function 
Now the coverage should be better 

Tomorrow I'll work on @patmmccann 's comments which will have the effect of removing the `SUPPORTED_BIDDERS` limitation

patmmccann commented on modules/1plusXRtdProvider.js:
```diff
@@ -0,0 +1,211 @@
+    name: '1plusX.com',
+    segment: segments.map((segmentId) => ({ id: segmentId }))
+  };
+  const site = {
+    keywords: topics.join(',')
+  };
```
you may notice that appnexus doesnt support site.content.data.segment.id. In that case, we would prefer a change to the appnexus adapter to put the data in the keywords vector than a change to your module to write it to where it no longer belongs or special appnexus handling in your adapter. We have an open issue to remove similar special handling for appnexus in the airgrid and permutive adapters (among possible others)

patmmccann replied:
#8596

bwhisp commented on integrationExamples/gpt/1plusXRtdProviderExample.html:
```diff
@@ -0,0 +1,112 @@
+      gads.type = 'text/javascript';
+      var useSSL = 'https:' == document.location.protocol;
+      gads.src = (useSSL ? 'https:' : 'http:') +
+        '//www.googletagservices.com/tag/js/gpt.js';
```
Hi @patmmccann 
I took this piece of code from another module implementation example. 
These same lines are used in several module examples. 

Shall I take `/workspaces/Prebid.js/integrationExamples/gpt/hello_world.html` as a model instead ?

patmmccann replied:
We have an open docs issue to update all the locations to gpt.js new official host (doubleclick.net), while that is still outstanding, we'd prefer to not add things to fix later

bwhisp replied:
Done :)

bwhisp commented:
Hi @patmmccann 

thank you for the review

I've worked on the requested changes:
* removed the supported bidders limitation
* changed location to write topics to `site.content.data.segment.id`
* introduced legacy processing for `appnexus` (writing data to `site.keywords` for this bidder)
* updated tests

patmmccann commented:
@bwhisp thank you for promptly addressing the rules concerns.

@Fawke remains the code reviewer; i just wanted to assist with rules feedback

bwhisp commented on modules/1plusXRtdProvider.js:
```diff
@@ -0,0 +1,235 @@
+  } else {
+    const siteContentData = {
+      name: ORTB2_NAME,
+      segment: topics.map((topicId) => ({ id: topicId }))
```
I will take care of it tomorrow

bwhisp replied:
shall I also write a segtax for what I put in `user.data.segment` ?

Fawke commented:
@bwhisp Yeah, you can pull in master and perform the merge.

bwhisp commented:
I messed up...

bwhisp commented:
Hi @Fawke 

following your feedback I updated the handling of Appnexus & I resolved the merge conflict

bwhisp commented:
Hello @patmmccann @Fawke 
I created a PR at https://github.com/InteractiveAdvertisingBureau/openrtb/pull/108 to introduce the 1plusX audience & content taxonomies that will be used in this RTD module 
Do you know if it's a prerequisite that https://github.com/InteractiveAdvertisingBureau/openrtb/pull/108 is merged before #8614?
If yes, do you know who can perform a review? I've seen @patmmccann doing a few of them in the past but not 100% sure it's still the case

patmmccann commented:
IAB reviews those
